<a href="https://colab.research.google.com/github/akshayabhale1406/SCLR_Delhivery_FE/blob/main/Delhivery_FE.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Business Case: Delhivery - Feature Engineering**

## Problem Statement : About Delhivery

Delhivery is the largest and fastest-growing fully integrated player in India by revenue in Fiscal 2021. They aim to build the operating system for commerce, through a combination of world-class infrastructure, logistics operations of the highest quality, and cutting-edge engineering and technology capabilities.

The Data team builds intelligence and capabilities using this data that helps them to widen the gap between the quality, efficiency, and profitability of their business versus their competitors.

 **Approach to understanding the problem and getting to insights**

First we will understand the data we have received, and then check what possible operation we can do to analyse the data better and observe patterns to provide solutions.

### Importing the relevant libraries

In [128]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns
pd.set_option('display.max_columns', None)

### Loading CSV as Dataset

In [129]:
df = pd.read_csv("/content/sample_data/delhivery_data.csv")



### Checking if the data is loaded as dataframe

In [130]:
df.head()

data          trip_creation_time  \
0  training  2018-09-20 02:35:36.476840   
1  training  2018-09-20 02:35:36.476840   
2  training  2018-09-20 02:35:36.476840   
3  training  2018-09-20 02:35:36.476840   
4  training  2018-09-20 02:35:36.476840   

                                 route_schedule_uuid route_type  \
0  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
1  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
2  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
3  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
4  thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   

                 trip_uuid source_center                 source_name  \
0  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
1  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
2  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
3  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
4  trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   

  destination_center               destination_name  \
0       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
1       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
2       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
3       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
4       IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   

                od_start_time                 od_end_time  \
0  2018-09-20 03:21:32.418600  2018-09-20 04:47:45.236797   
1  2018-09-20 03:21:32.418600  2018-09-20 04:47:45.236797   
2  2018-09-20 03:21:32.418600  2018-09-20 04:47:45.236797   
3  2018-09-20 03:21:32.418600  2018-09-20 04:47:45.236797   
4  2018-09-20 03:21:32.418600  2018-09-20 04:47:45.236797   

   start_scan_to_end_scan  is_cutoff  cutoff_factor  \
0                    86.0       True              9   
1                    86.0       True             18   
2                    86.0       True             27   
3                    86.0       True             36   
4                    86.0      False             39   

             cutoff_timestamp  actual_distance_to_destination  actual_time  \
0         2018-09-20 04:27:55                       10.435660         14.0   
1         2018-09-20 04:17:55                       18.936842         24.0   
2  2018-09-20 04:01:19.505586                       27.637279         40.0   
3         2018-09-20 03:39:57                       36.118028         62.0   
4         2018-09-20 03:33:55                       39.386040         68.0   

   osrm_time  osrm_distance    factor  segment_actual_time  segment_osrm_time  \
0       11.0        11.9653  1.272727                 14.0               11.0   
1       20.0        21.7243  1.200000                 10.0                9.0   
2       28.0        32.5395  1.428571                 16.0                7.0   
3       40.0        45.5620  1.550000                 21.0               12.0   
4       44.0        54.2181  1.545455                  6.0                5.0   

   segment_osrm_distance  segment_factor  
0                11.9653        1.272727  
1                 9.7590        1.111111  
2                10.8152        2.285714  
3                13.0224        1.750000  
4                 3.9153        1.200000

### Checking the shape of the data

In [131]:
df.shape

(144867, 24)

### Checking column name

In [132]:
df.columns

Index(['data', 'trip_creation_time', 'route_schedule_uuid', 'route_type',
       'trip_uuid', 'source_center', 'source_name', 'destination_center',
       'destination_name', 'od_start_time', 'od_end_time',
       'start_scan_to_end_scan', 'is_cutoff', 'cutoff_factor',
       'cutoff_timestamp', 'actual_distance_to_destination', 'actual_time',
       'osrm_time', 'osrm_distance', 'factor', 'segment_actual_time',
       'segment_osrm_time', 'segment_osrm_distance', 'segment_factor'],
      dtype='object')

**OBSERVATION**: This data consists of 144867 checkpoints described across 24 columns.


### Checking data type of each column

In [133]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144867 entries, 0 to 144866
Data columns (total 24 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   data                            144867 non-null  object 
 1   trip_creation_time              144867 non-null  object 
 2   route_schedule_uuid             144867 non-null  object 
 3   route_type                      144867 non-null  object 
 4   trip_uuid                       144867 non-null  object 
 5   source_center                   144867 non-null  object 
 6   source_name                     144574 non-null  object 
 7   destination_center              144867 non-null  object 
 8   destination_name                144606 non-null  object 
 9   od_start_time                   144867 non-null  object 
 10  od_end_time                     144867 non-null  object 
 11  start_scan_to_end_scan          144867 non-null  float64
 12  is_cutoff       

**OBSERVATION**: The time column should be of correct datatype, that is: datetime

### Convert the od_start_time and od_end_time to datetime data type

In [134]:
df["trip_creation_time"] = pd.to_datetime(df["trip_creation_time"])
df["od_start_time"] = pd.to_datetime(df["od_start_time"])
df["od_end_time"] = pd.to_datetime(df["od_end_time"])

In [135]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144867 entries, 0 to 144866
Data columns (total 24 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   data                            144867 non-null  object        
 1   trip_creation_time              144867 non-null  datetime64[ns]
 2   route_schedule_uuid             144867 non-null  object        
 3   route_type                      144867 non-null  object        
 4   trip_uuid                       144867 non-null  object        
 5   source_center                   144867 non-null  object        
 6   source_name                     144574 non-null  object        
 7   destination_center              144867 non-null  object        
 8   destination_name                144606 non-null  object        
 9   od_start_time                   144867 non-null  datetime64[ns]
 10  od_end_time                     144867 non-null  datetim

### Checking for missing vaules in column

In [136]:
df.isnull().sum()

data                                0
trip_creation_time                  0
route_schedule_uuid                 0
route_type                          0
trip_uuid                           0
source_center                       0
source_name                       293
destination_center                  0
destination_name                  261
od_start_time                       0
od_end_time                         0
start_scan_to_end_scan              0
is_cutoff                           0
cutoff_factor                       0
cutoff_timestamp                    0
actual_distance_to_destination      0
actual_time                         0
osrm_time                           0
osrm_distance                       0
factor                              0
segment_actual_time                 0
segment_osrm_time                   0
segment_osrm_distance               0
segment_factor                      0
dtype: int64

**OBSERVATION**: There are missing values in  columns: source_name, destination_name, and segment factor. Since the missing values are less than 1% of the total data, we can directly drop the NULL values.



### Dropping the NULL values from the dataset

In [137]:
df = df.dropna(how ='any')

### Checking the shape of data after dropping NULL Values

In [138]:
df.shape

(144316, 24)

**OBSERVATION**: The new dataset is decreased by less than 1% of the actual raw dataset. This is acceptable for further Data Analysis.

In [139]:
df_null_check = df.isnull().sum()

### Checking to see if there are any NULL values in the new dataset.

In [140]:
df_null_check

data                              0
trip_creation_time                0
route_schedule_uuid               0
route_type                        0
trip_uuid                         0
source_center                     0
source_name                       0
destination_center                0
destination_name                  0
od_start_time                     0
od_end_time                       0
start_scan_to_end_scan            0
is_cutoff                         0
cutoff_factor                     0
cutoff_timestamp                  0
actual_distance_to_destination    0
actual_time                       0
osrm_time                         0
osrm_distance                     0
factor                            0
segment_actual_time               0
segment_osrm_time                 0
segment_osrm_distance             0
segment_factor                    0
dtype: int64

**OBSERVATION**: No NULL values in the new dataset.

### Checking the statistical summary of raw data

In [141]:
df.describe()

trip_creation_time                  od_start_time  \
count                         144316                         144316   
mean   2018-09-22 13:05:09.454117120  2018-09-22 17:32:42.435769344   
min       2018-09-12 00:00:16.535741     2018-09-12 00:00:16.535741   
25%    2018-09-17 02:46:11.004421120  2018-09-17 07:37:35.014584832   
50%    2018-09-22 03:36:19.186585088  2018-09-22 07:35:23.038482944   
75%    2018-09-27 17:53:19.027942912  2018-09-27 22:01:30.861209088   
max       2018-10-03 23:59:42.701692     2018-10-06 04:27:23.392375   
std                              NaN                            NaN   

                         od_end_time  start_scan_to_end_scan  cutoff_factor  \
count                         144316           144316.000000  144316.000000   
mean   2018-09-23 09:36:54.057172224              963.697698     233.561345   
min       2018-09-12 00:50:10.814399               20.000000       9.000000   
25%       2018-09-18 01:29:56.978912              161.000000      22.000000   
50%    2018-09-23 02:49:00.936600064              451.000000      66.000000   
75%    2018-09-28 12:13:41.675546112             1645.000000     286.000000   
max       2018-10-08 03:00:24.353479             7898.000000    1927.000000   
std                              NaN             1038.082976     345.245823   

       actual_distance_to_destination    actual_time      osrm_time  \
count                   144316.000000  144316.000000  144316.000000   
mean                       234.708498     417.996237     214.437055   
min                          9.000045       9.000000       6.000000   
25%                         23.352027      51.000000      27.000000   
50%                         66.135322     132.000000      64.000000   
75%                        286.919294     516.000000     259.000000   
max                       1927.447705    4532.000000    1686.000000   
std                        345.480571     598.940065     308.448543   

       osrm_distance         factor  segment_actual_time  segment_osrm_time  \
count  144316.000000  144316.000000        144316.000000      144316.000000   
mean      285.549785       2.120178            36.175379          18.495697   
min         9.008200       0.144000          -244.000000           0.000000   
25%        29.896250       1.604545            20.000000          11.000000   
50%        78.624400       1.857143            28.000000          17.000000   
75%       346.305400       2.212280            40.000000          22.000000   
max      2326.199100      77.387097          3051.000000        1611.000000   
std       421.717826       1.717065            53.524298          14.774008   

       segment_osrm_distance  segment_factor  
count          144316.000000   144316.000000  
mean               22.818993        2.218707  
min                 0.000000      -23.444444  
25%                12.053975        1.347826  
50%                23.508300        1.684211  
75%                27.813325        2.250000  
max              2191.403700      574.250000  
std                17.866367        4.854804

### Checking if any orders are returned to source location.

In [142]:
returned_orders = df.groupby("trip_uuid").apply(lambda group: pd.DataFrame({"Returned_Packages": [any(group["source_center"] == group["destination_center"])]})).reset_index()
returned_orders.head()

<ipython-input-142-887a70d66ffd>:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  returned_orders = df.groupby("trip_uuid").apply(lambda group: pd.DataFrame({"Returned_Packages": [any(group["source_center"] == group["destination_center"])]})).reset_index()


trip_uuid  level_1  Returned_Packages
0  trip-153671041653548748        0              False
1  trip-153671042288605164        0              False
2  trip-153671043369099517        0              False
3  trip-153671046011330457        0              False
4  trip-153671052974046625        0              False

In [143]:
true_instances = returned_orders[returned_orders['Returned_Packages'] == True].index.tolist()
true_instances

[]

**OBSERVATION**: The below shows if any values are TRUE for Returned_Packages. Since the returned list is empty, it means no packages were returned.

### The current details are splitted for different trips as per different check-points for that trips. Data can be analysed better if the times of each check-points can be summed up for each trip ID.

In [144]:
df_combined_times = df.groupby("trip_uuid")[["segment_actual_time", "segment_osrm_time", "segment_osrm_distance"]].sum()
df_combined_times["segment_osrm_distance"] = np.round(df_combined_times["segment_osrm_distance"], 2)
df_combined_times = df_combined_times.rename(columns = {"segment_osrm_time" : "accumulated_osrm_time", "segment_actual_time" : "accumulated_actual_time", "segment_osrm_distance" : "accumulated_osrm_distance"})
df_combined_times.head()

accumulated_actual_time  accumulated_osrm_time  \
trip_uuid                                                                 
trip-153671041653548748                   1548.0                 1008.0   
trip-153671042288605164                    141.0                   65.0   
trip-153671043369099517                   3308.0                 1941.0   
trip-153671046011330457                     59.0                   16.0   
trip-153671052974046625                    340.0                  115.0   

                         accumulated_osrm_distance  
trip_uuid                                           
trip-153671041653548748                    1320.47  
trip-153671042288605164                      84.19  
trip-153671043369099517                    2545.27  
trip-153671046011330457                      19.88  
trip-153671052974046625                     146.79

In [145]:
df_combined_times_1 = df.groupby(["trip_uuid","destination_center"])[["actual_time", "osrm_time", "osrm_distance"]].last()
df_combined_times_1 = df_combined_times_1.reset_index()
df_combined_times_1 = df_combined_times_1.groupby("trip_uuid")[["actual_time", "osrm_time", "osrm_distance"]].sum()
df_combined_times_1.head()

actual_time  osrm_time  osrm_distance
trip_uuid                                                     
trip-153671041653548748       1562.0      717.0       991.3523
trip-153671042288605164        143.0       68.0        85.1110
trip-153671043369099517       3347.0     1740.0      2354.0665
trip-153671046011330457         59.0       15.0        19.6800
trip-153671052974046625        341.0      117.0       146.7918

### Calculation of time difference between OSRM time and Actual Time to know how much time error present present in the OSRM time calculation.

In [146]:
df_combined_times_1 = df_combined_times_1.reset_index()
df_combined_times = pd.merge(df_combined_times, df_combined_times_1, on = "trip_uuid")
df_combined_times.head()

trip_uuid  accumulated_actual_time  accumulated_osrm_time  \
0  trip-153671041653548748                   1548.0                 1008.0   
1  trip-153671042288605164                    141.0                   65.0   
2  trip-153671043369099517                   3308.0                 1941.0   
3  trip-153671046011330457                     59.0                   16.0   
4  trip-153671052974046625                    340.0                  115.0   

   accumulated_osrm_distance  actual_time  osrm_time  osrm_distance  
0                    1320.47       1562.0      717.0       991.3523  
1                      84.19        143.0       68.0        85.1110  
2                    2545.27       3347.0     1740.0      2354.0665  
3                      19.88         59.0       15.0        19.6800  
4                     146.79        341.0      117.0       146.7918

In [147]:
df

data         trip_creation_time  \
0       training 2018-09-20 02:35:36.476840   
1       training 2018-09-20 02:35:36.476840   
2       training 2018-09-20 02:35:36.476840   
3       training 2018-09-20 02:35:36.476840   
4       training 2018-09-20 02:35:36.476840   
...          ...                        ...   
144862  training 2018-09-20 16:24:28.436231   
144863  training 2018-09-20 16:24:28.436231   
144864  training 2018-09-20 16:24:28.436231   
144865  training 2018-09-20 16:24:28.436231   
144866  training 2018-09-20 16:24:28.436231   

                                      route_schedule_uuid route_type  \
0       thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
1       thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
2       thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
3       thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
4       thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...    Carting   
...                                                   ...        ...   
144862  thanos::sroute:f0569d2f-4e20-4c31-8542-67b86d5...    Carting   
144863  thanos::sroute:f0569d2f-4e20-4c31-8542-67b86d5...    Carting   
144864  thanos::sroute:f0569d2f-4e20-4c31-8542-67b86d5...    Carting   
144865  thanos::sroute:f0569d2f-4e20-4c31-8542-67b86d5...    Carting   
144866  thanos::sroute:f0569d2f-4e20-4c31-8542-67b86d5...    Carting   

                      trip_uuid source_center                 source_name  \
0       trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
1       trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
2       trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
3       trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
4       trip-153741093647649320  IND388121AAA  Anand_VUNagar_DC (Gujarat)   
...                         ...           ...                         ...   
144862  trip-153746066843555182  IND131028AAB  Sonipat_Kundli_H (Haryana)   
144863  trip-153746066843555182  IND131028AAB  Sonipat_Kundli_H (Haryana)   
144864  trip-153746066843555182  IND131028AAB  Sonipat_Kundli_H (Haryana)   
144865  trip-153746066843555182  IND131028AAB  Sonipat_Kundli_H (Haryana)   
144866  trip-153746066843555182  IND131028AAB  Sonipat_Kundli_H (Haryana)   

       destination_center               destination_name  \
0            IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
1            IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
2            IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
3            IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
4            IND388620AAB  Khambhat_MotvdDPP_D (Gujarat)   
...                   ...                            ...   
144862       IND000000ACB  Gurgaon_Bilaspur_HB (Haryana)   
144863       IND000000ACB  Gurgaon_Bilaspur_HB (Haryana)   
144864       IND000000ACB  Gurgaon_Bilaspur_HB (Haryana)   
144865       IND000000ACB  Gurgaon_Bilaspur_HB (Haryana)   
144866       IND000000ACB  Gurgaon_Bilaspur_HB (Haryana)   

                    od_start_time                od_end_time  \
0      2018-09-20 03:21:32.418600 2018-09-20 04:47:45.236797   
1      2018-09-20 03:21:32.418600 2018-09-20 04:47:45.236797   
2      2018-09-20 03:21:32.418600 2018-09-20 04:47:45.236797   
3      2018-09-20 03:21:32.418600 2018-09-20 04:47:45.236797   
4      2018-09-20 03:21:32.418600 2018-09-20 04:47:45.236797   
...                           ...                        ...   
144862 2018-09-20 16:24:28.436231 2018-09-20 23:32:09.618069   
144863 2018-09-20 16:24:28.436231 2018-09-20 23:32:09.618069   
144864 2018-09-20 16:24:28.436231 2018-09-20 23:32:09.618069   
144865 2018-09-20 16:24:28.436231 2018-09-20 23:32:09.618069   
144866 2018-09-20 16:24:28.436231 2018-09-20 23:32:09.618069   

        start_scan_to_end_scan  is_cutoff  cutoff_factor  \
0                         86.0       True              9   
1                         86.0       True      

In [148]:
od_start_time = df.groupby('trip_uuid')['od_start_time'].first()
od_start_time = od_start_time.reset_index()
od_start_time = od_start_time.rename(columns={'od_start_time': 'trip_start_time'})
df_combined_times = df_combined_times.merge(od_start_time, on='trip_uuid', how='left').reset_index(drop=True)
# The error is likely on the line below. Make sure "trip_start_time" column exists
df_combined_times["trip_start_date"] = df_combined_times["trip_start_time"].dt.date
trip_start_location = df.groupby('trip_uuid')['source_name'].first()
trip_end_location = df.groupby('trip_uuid')['destination_name'].last()
trip_end_location = trip_end_location.reset_index()
trip_start_location = trip_start_location.reset_index()
df_combined_times = df_combined_times.merge(trip_start_location, on='trip_uuid', how='left').reset_index(drop=True)
df_combined_times = df_combined_times.merge(trip_end_location, on='trip_uuid', how='left').reset_index(drop=True)
df_combined_times = df_combined_times.drop(columns=['trip_start_time'])
df_combined_times.head()

trip_uuid  accumulated_actual_time  accumulated_osrm_time  \
0  trip-153671041653548748                   1548.0                 1008.0   
1  trip-153671042288605164                    141.0                   65.0   
2  trip-153671043369099517                   3308.0                 1941.0   
3  trip-153671046011330457                     59.0                   16.0   
4  trip-153671052974046625                    340.0                  115.0   

   accumulated_osrm_distance  actual_time  osrm_time  osrm_distance  \
0                    1320.47       1562.0      717.0       991.3523   
1                      84.19        143.0       68.0        85.1110   
2                    2545.27       3347.0     1740.0      2354.0665   
3                      19.88         59.0       15.0        19.6800   
4                     146.79        341.0      117.0       146.7918   

  trip_start_date                         source_name  \
0      2018-09-12  Bhopal_Trnsport_H (Madhya Pradesh)   
1      2018-09-12       Tumkur_Veersagr_I (Karnataka)   
2      2018-09-12    Bangalore_Nelmngla_H (Karnataka)   
3      2018-09-12            Mumbai Hub (Maharashtra)   
4      2018-09-12              Bellary_Dc (Karnataka)   

                    destination_name  
0      Gurgaon_Bilaspur_HB (Haryana)  
1  Chikblapur_ShntiSgr_D (Karnataka)  
2     Chandigarh_Mehmdpur_H (Punjab)  
3     Mumbai_MiraRd_IP (Maharashtra)  
4             Bellary_Dc (Karnataka)

In [103]:
od_start_time = df.groupby('trip_uuid')['od_start_time'].first()
od_start_time = od_start_time.reset_index()
od_start_time = od_start_time.rename(columns={'od_start_time': 'trip_start_time'})
df_combined_times = df_combined_times.merge(od_start_time, on='trip_uuid', how='left').reset_index(drop=True)
df_combined_times["trip_start_date"] = df_combined_times["trip_start_time"].dt.date
trip_start_location = df.groupby('trip_uuid')['source_name'].first()
trip_end_location = df.groupby('trip_uuid')['destination_name'].last()
trip_end_location = trip_end_location.reset_index()
trip_start_location = trip_start_location.reset_index()
df_combined_times = df_combined_times.merge(trip_start_location, on='trip_uuid', how='left').reset_index()
df_combined_times = df_combined_times.merge(trip_end_location, on='trip_uuid', how='left').reset_index()
# df_combined_times = df_combined_times.drop(columns=['od_start_time'])
df_combined_times.head()

KeyError: 'trip_start_time'

In [ ]:
df_combined_times = df_combined_times.merge(od_start_time, on='trip_uuid', how='left').reset_index(drop=True)

### Creating a column for Starting Point and Ending Point of a trip using States names

In [149]:
df_combined_times["source_name"] = df_combined_times["source_name"].str.split("(").str[1]
df_combined_times["source_name"] = df_combined_times["source_name"].str.split(")").str[0]
df_combined_times["source_name"] = df_combined_times["source_name"].apply(lambda x : x.upper())
df_combined_times["destination_name"] = df_combined_times["destination_name"].str.split("(").str[1]
df_combined_times["destination_name"] = df_combined_times["destination_name"].str.split(")").str[0]
df_combined_times["destination_name"] = df_combined_times["destination_name"].apply(lambda x : x.upper())
df_combined_times

trip_uuid  accumulated_actual_time  \
0      trip-153671041653548748                   1548.0   
1      trip-153671042288605164                    141.0   
2      trip-153671043369099517                   3308.0   
3      trip-153671046011330457                     59.0   
4      trip-153671052974046625                    340.0   
...                        ...                      ...   
14782  trip-153861095625827784                     82.0   
14783  trip-153861104386292051                     21.0   
14784  trip-153861106442901555                    281.0   
14785  trip-153861115439069069                    258.0   
14786  trip-153861118270144424                    274.0   

       accumulated_osrm_time  accumulated_osrm_distance  actual_time  \
0                     1008.0                    1320.47       1562.0   
1                       65.0                      84.19        143.0   
2                     1941.0                    2545.27       3347.0   
3                       16.0                      19.88         59.0   
4                      115.0                     146.79        341.0   
...                      ...                        ...          ...   
14782                   62.0                      64.86         83.0   
14783                   11.0                      16.09         21.0   
14784                   88.0                     104.89        282.0   
14785                  221.0                     223.53        264.0   
14786                   67.0                      80.58        275.0   

       osrm_time  osrm_distance trip_start_date     source_name  \
0          717.0       991.3523      2018-09-12  MADHYA PRADESH   
1           68.0        85.1110      2018-09-12       KARNATAKA   
2         1740.0      2354.0665      2018-09-12       KARNATAKA   
3           15.0        19.6800      2018-09-12     MAHARASHTRA   
4          117.0       146.7918      2018-09-12       KARNATAKA   
...          ...            ...             ...             ...   
14782       62.0        73.4630      2018-10-03          PUNJAB   
14783       12.0        16.0882      2018-10-03         HARYANA   
14784       48.0        58.9037      2018-10-03   UTTAR PRADESH   
14785      179.0       171.1103      2018-10-03      TAMIL NADU   
14786       68.0        80.5787      2018-10-04       KARNATAKA   

      destination_name  
0              HARYANA  
1            KARNATAKA  
2               PUNJAB  
3          MAHARASHTRA  
4            KARNATAKA  
...                ...  
14782           PUNJAB  
14783          HARYANA  
14784    UTTAR PRADESH  
14785       TAMIL NADU  
14786        KARNATAKA  

[14787 rows x 10 columns]

In [150]:
od_start = df.groupby('trip_uuid')['od_start_time'].first()
od_end = df.groupby('trip_uuid')['od_end_time'].last()
df_combined_times = df_combined_times.set_index('trip_uuid').merge(od_start, left_index=True, right_index=True).reset_index()
df_combined_times = df_combined_times.set_index('trip_uuid').merge(od_end, left_index=True, right_index=True).reset_index()
df_combined_times["trip_duration"] = (df_combined_times["od_end_time"] - df_combined_times["od_start_time"]).dt.total_seconds()/60
df_combined_times.head()

trip_uuid  accumulated_actual_time  accumulated_osrm_time  \
0  trip-153671041653548748                   1548.0                 1008.0   
1  trip-153671042288605164                    141.0                   65.0   
2  trip-153671043369099517                   3308.0                 1941.0   
3  trip-153671046011330457                     59.0                   16.0   
4  trip-153671052974046625                    340.0                  115.0   

   accumulated_osrm_distance  actual_time  osrm_time  osrm_distance  \
0                    1320.47       1562.0      717.0       991.3523   
1                      84.19        143.0       68.0        85.1110   
2                    2545.27       3347.0     1740.0      2354.0665   
3                      19.88         59.0       15.0        19.6800   
4                     146.79        341.0      117.0       146.7918   

  trip_start_date     source_name destination_name              od_start_time  \
0      2018-09-12  MADHYA PRADESH          HARYANA 2018-09-12 00:00:16.535741   
1      2018-09-12       KARNATAKA        KARNATAKA 2018-09-12 00:00:22.886430   
2      2018-09-12       KARNATAKA           PUNJAB 2018-09-12 00:00:33.691250   
3      2018-09-12     MAHARASHTRA      MAHARASHTRA 2018-09-12 00:01:00.113710   
4      2018-09-12       KARNATAKA        KARNATAKA 2018-09-12 00:02:09.740725   

                 od_end_time  trip_duration  
0 2018-09-13 13:40:23.123744    2260.109800  
1 2018-09-12 03:01:59.598855     181.611874  
2 2018-09-14 17:34:55.442454    3934.362520  
3 2018-09-12 01:41:29.809822     100.494935  
4 2018-09-12 12:00:30.683231     718.349042

In [ ]:
df_combined_times.describe()

In [ ]:
df_combined_times = df_combined_times.drop(columns=['od_start_time', 'od_end_time'])
start_scan_to_end_scan = df.groupby(["trip_uuid","destination_center"])["start_scan_to_end_scan"].first()
start_scan_to_end_scan = start_scan_to_end_scan.reset_index()
start_scan_to_end_scan.head()

In [ ]:
df_combined_times1 = start_scan_to_end_scan.groupby("trip_uuid")["start_scan_to_end_scan"].sum()
df_combined_times1 = df_combined_times1.reset_index()
df_combined_times1.head()

In [ ]:
df_combined_times = df_combined_times.merge(df_combined_times1, on = "trip_uuid", how = "left")
df_combined_times.head()

In [ ]:
df_route_type = df.groupby("trip_uuid")["route_type"].first()
df_route_type.reset_index()
df_combined_times = df_combined_times.merge(df_route_type, left_on = "trip_uuid", right_on = "trip_uuid")
# df_combined_times = df_combined_times.rename(columns = {'segment_osrm_time' : "osrm" })
df_combined_times

In [ ]:
df_top_10_center = df["source_center"].value_counts().head(13)

df_top_10_destination_center = df["destination_center"].value_counts().head(13)

df_most_busy_center = df_top_10_center.reset_index().merge(df_top_10_destination_center.reset_index(), left_on="source_center", right_on="destination_center")

df_most_busy_center = df_most_busy_center.rename(columns={"source_center": "Center_Code"})

df_most_busy_center = df_most_busy_center.drop(columns=["destination_center"])

df_most_busy_center = df_most_busy_center.rename(columns = {"count_x" : "source_center", "count_y" : "destination_center"})

df_most_busy_center

In [ ]:
df_most_busy_center.plot(x = "Center_Code", y=["source_center", "destination_center"], kind="bar")

In [ ]:
df_top_10_city = df_combined_times["source_name"].value_counts().head(10)

df_top_10_destination_name = df_combined_times["destination_name"].value_counts().head(10)

df_most_busy_city = df_top_10_city.reset_index().merge(df_top_10_destination_name.reset_index(), left_on="source_name", right_on="destination_name")

df_most_busy_city = df_most_busy_city.rename(columns={"source_name": "city_name"})

df_most_busy_city = df_most_busy_city.drop(columns=["destination_name"])

df_most_busy_city = df_most_busy_city.rename(columns = {"count_x" : "from_city", "count_y" : "to_city"})

df_most_busy_city

In [ ]:
df_most_busy_city.plot(x = "city_name", y=["from_city", "to_city"], kind="bar")

In [ ]:
df_bottom_10_city = df_combined_times["source_name"].value_counts().tail(13)

df_bottom_10_destination_name = df_combined_times["destination_name"].value_counts().tail(13)

df_least_busy_city = df_bottom_10_city.reset_index().merge(df_bottom_10_destination_name.reset_index(), left_on="source_name", right_on="destination_name")

df_least_busy_city = df_least_busy_city.rename(columns={"source_name": "city_name"})

df_least_busy_city = df_least_busy_city.drop(columns=["destination_name"])

df_least_busy_city = df_least_busy_city.rename(columns = {"count_x" : "from_city", "count_y" : "to_city"})

df_least_busy_city

In [ ]:
df_least_busy_city.plot(x = "city_name", y=["from_city", "to_city"], kind="bar")

In [ ]:
df_bottom_10_city = df_combined_times["source_name"].value_counts().tail(10)
df_bottom_10_city

## HYPOTHESIS TESTING:

1) Testing if the difference trip_duration and start_scan_to_end_scan is significant using **TWO SAMPLE T-TEST**

In [ ]:
df_hypo_test1 = df_combined_times.copy()
df_hypo_test1["diff_trip_duration_scan"] = df_hypo_test1["trip_duration"] - df_hypo_test1["start_scan_to_end_scan"]
df_hypo_test1.head()

In [ ]:
df_hypo_test1.describe()

In [ ]:
import matplotlib.pyplot as plt
plt.figure(figsize=(10, 6))
plt.hist(df_hypo_test1["diff_trip_duration_scan"], bins=50, edgecolor='black')
plt.xlabel("Difference (Trip Duration - Scan Time)")
plt.ylabel("Frequency")
plt.title("Histogram of Difference between Trip Duration and Scan Time")
plt.grid(True)
plt.show()

In [ ]:
def remove_outliers_iqr_df(df_data, column_name):

  q1 = df_data[column_name].quantile(0.25)
  q3 = df_data[column_name].quantile(0.75)
  iqr = q3 - q1

  lower_bound = q1 - 1.5 * iqr
  upper_bound = q3 + 1.5 * iqr

  return df_data[(df_data[column_name] >= lower_bound) & (df_data[column_name] <= upper_bound)]

data_without_outliers = remove_outliers_iqr_df(df_hypo_test1, 'start_scan_to_end_scan')

In [ ]:
def remove_outliers_iqr_df1(df_data1, column_name1):

  q1 = df_data1[column_name].quantile(0.25)
  q3 = df_data1[column_name].quantile(0.75)
  iqr = q3 - q1

  lower_bound = q1 - 1.5 * iqr
  upper_bound = q3 + 1.5 * iqr

  return df_data1[(df_data1[column_name] >= lower_bound) & (df_data1[column_name] <= upper_bound)]

data_without_outliers1 = remove_outliers_iqr_df(data_without_outliers, 'trip_duration')

In [ ]:
data_without_outliers1.describe()

In [ ]:
from scipy import stats
t_statistic, p_value = stats.ttest_rel(df_hypo_test1["trip_duration"], df_hypo_test1["start_scan_to_end_scan"])
print("t-statistic:", t_statistic)
print("p-value:", p_value)

alpha = 0.05
if p_value < alpha:
    print("Reject the null hypothesis. There is a significant difference between the means of trip_duration and start_scan_to_end_scan.")
else:
    print("Fail to reject the null hypothesis. No significant difference between the means of trip_duration and start_scan_to_end_scan.")

In [ ]:
data_without_outliers1

In [ ]:
df_melted = pd.melt(data_without_outliers1.reset_index(), id_vars=['trip_uuid'], value_vars=['trip_duration', 'start_scan_to_end_scan'])
sns.boxplot(x='variable', y='value', data=df_melted)
plt.title('Box Plot')
plt.show()

actual_time aggregated value and OSRM time aggregated value


In [ ]:
df_act_osrm_test = df.groupby(["trip_uuid", "destination_center"] )[["actual_time", "osrm_time"]].last()
df_act_osrm_test = df_act_osrm_test.reset_index()
df_act_osrm_test = df_act_osrm_test.groupby("trip_uuid")[["actual_time", "osrm_time"]].sum()
df_act_osrm_test

In [ ]:
df_act_osrm_test.describe()

In [ ]:
df_melted_1 = pd.melt(df_act_osrm_test.reset_index(), id_vars = ["trip_uuid"], value_vars = ["actual_time", "osrm_time"])

sns.boxplot(x = "variable", y = "value", data = df_melted_1)

plt.title("Box plot for actual_time aggregated value and OSRM time aggregated value")
plt.show()

In [ ]:
def remove_outliers_iqr_df_act_osrm_test(df_act, act_t, osrm_t):

  q1 = df_act[act_t].quantile(0.25)
  q3 = df_act[act_t].quantile(0.75)
  iqr1 = q3 - q1

  lower_bound1 = q1 - 1.5 * iqr1
  upper_bound2 = q3 + 1.5 * iqr1

  q2 = df_act[osrm_t].quantile(0.25)
  q4 = df_act[osrm_t].quantile(0.75)
  iqr2 = q4 - q2

  lower_bound3 = q2 - 1.5 * iqr2
  upper_bound4 = q4 + 1.5 * iqr2

  return df_act[((df_act[act_t] >= lower_bound1) & (df_act[act_t] <= upper_bound2) & (df_act[osrm_t] >= lower_bound3) & (df_act[osrm_t] <= upper_bound4))]

df_act_osrm_test_wo_outliers = remove_outliers_iqr_df_act_osrm_test(df_act_osrm_test, 'actual_time', 'osrm_time')

In [ ]:
df_act_osrm_test_wo_outliers

In [ ]:
df_act_osrm_test_wo_outliers.describe()

In [ ]:
df_melted_2 = pd.melt(df_act_osrm_test_wo_outliers.reset_index(), id_vars = ["trip_uuid"], value_vars = ["actual_time", "osrm_time"])

sns.boxplot(x = "variable", y = "value", data = df_melted_2)

plt.title("Box plot for actual_time aggregated value and OSRM time aggregated value")
plt.show()

### actual_time aggregated value and segment actual time aggregated value

In [ ]:
df_atav = df.groupby(["trip_uuid", "destination_center"] )[["actual_time"]].last()
df_atav = df_atav.reset_index()
df_atav = df_atav.groupby("trip_uuid")["actual_time"].sum()
df_atav

In [ ]:
df_atav_seg = df.groupby(["trip_uuid"])["segment_actual_time"].sum()
df_atav = df_atav.reset_index()
df_atav = df_atav.merge(df_atav_seg, left_on='trip_uuid', right_on='trip_uuid')
df_atav = df_atav.reset_index()
df_atav

In [ ]:
df_atav.describe()

In [ ]:
def remove_outliers_iqr_df_test3(df_atav_out, act_t, osrm_t):

  for col in df_atav_out.columns:
    q1 = df_atav_out[col].quantile(0.25)
    q2 = df_atav_out[col].quantile(0.75)
    iqr = q2 -q1
    lower_bound = q1 - 1.5 * iqr
    upper_bound = q2 + 1.5 * iqr
    df_atav_out = df_atav_out[(df_atav_out[col] >= lower_bound) & (df_atav_out[col] <= upper_bound)]

    return df_atav_out

df_atav_out = remove_outliers_iqr_df_test3(df_atav, "actual_time", "segment_actual_time")


In [ ]:
df_atav_out

In [ ]:
df_atav_out.describe()

In [ ]:
t_statistic, p_value = stats.ttest_rel(df_atav_out["actual_time"], df_atav_out["segment_actual_time"])
print("t-statistic:", t_statistic)
print("p-value:", p_value)

alpha = 0.05
if p_value < alpha:
    print("Reject the null hypothesis. There is a significant difference between the means of trip_duration and start_scan_to_end_scan.")
else:
    print("Fail to reject the null hypothesis. No significant difference between the means of trip_duration and start_scan_to_end_scan.")

In [ ]:
df_melted_2 = pd.melt(df_atav_out.reset_index(), id_vars = ["trip_uuid"], value_vars = ["actual_time", "segment_actual_time"])

sns.boxplot(x = "variable", y = "value", data = df_melted_2)

plt.title("Box plot for actual_time aggregated value and OSRM time aggregated value")
plt.show()

### osrm distance aggregated value and segment osrm distance aggregated value

In [ ]:
df_odav = df.groupby(["trip_uuid", "destination_center"] )[["osrm_distance"]].last()
df_odav = df_odav.reset_index()
df_odav = df_odav.groupby("trip_uuid")["osrm_distance"].sum()
df_odav

In [ ]:
df_odav_seg = df.groupby(["trip_uuid"])["segment_osrm_distance"].sum()
df_odav = df_odav.reset_index()
df_odav = df_odav.merge(df_odav_seg, left_on='trip_uuid', right_on='trip_uuid')
df_odav = df_odav.reset_index()
df_odav

In [ ]:
df_odav.describe()

In [ ]:
def df_odav_remove_outliers(odav, column):
    for col in columns:
      q1 = odav[col].quantile(0.25)
      q3 = odav[col].quantile(0.75)
      iqr = q3 - q1
      lower_bound = q1 - 1.5 * iqr
      upper_bound = q3 + 1.5 * iqr
      df_odav_clean = odav[(odav[col] >= lower_bound) & (odav[col] <= upper_bound)]

    return df_odav_clean

columns = ['osrm_distance', 'segment_osrm_distance']
df_odav_outremoved = df_odav_remove_outliers(df_odav, columns)
df_odav_outremoved

In [ ]:
df_odav_outremoved.describe()

In [ ]:
t_statistic, p_value = stats.ttest_rel(df_odav_outremoved["osrm_distance"], df_odav_outremoved["segment_osrm_distance"])
print("t-statistic:", t_statistic)
print("p-value:", p_value)

alpha = 0.05
if p_value < alpha:
    print("Reject the null hypothesis. There is a significant difference between the means of osrm_distance and segment_osrm_distance.")
else:
    print("Fail to reject the null hypothesis. No significant difference between the means of osrm_distance and segment_osrm_distance.")

In [ ]:
df_melted_3 = pd.melt(df_odav_outremoved.reset_index(), id_vars = ["trip_uuid"], value_vars = ["osrm_distance", "segment_osrm_distance"])

sns.boxplot(x = "variable", y = "value", data = df_melted_3)

plt.title("Box plot for  osrm distance aggregated value and segment osrm distance aggregated value ")
plt.show()

### hypothesis testing/ visual analysis between osrm time aggregated value and segment osrm time aggregated value

In [ ]:
df_otav = df.groupby(["trip_uuid", "destination_center"] )[["osrm_time"]].last()
df_otav = df_otav.reset_index()
df_otav = df_otav.groupby("trip_uuid")["osrm_time"].sum()

In [ ]:
df_otav_seg = df.groupby(["trip_uuid"])["segment_osrm_time"].sum()
df_otav = df_otav.reset_index()
df_otav = df_otav.merge(df_otav_seg, left_on='trip_uuid', right_on='trip_uuid')
df_otav = df_otav.reset_index()
df_otav

In [ ]:
df_otav.describe()

In [ ]:
def df_otav_remove_outliers(otav, column):
    for col in columns:
      q1 = otav[col].quantile(0.25)
      q3 = otav[col].quantile(0.75)
      iqr = q3 - q1
      lower_bound = q1 - 1.5 * iqr
      upper_bound = q3 + 1.5 * iqr
      df_otav_clean = otav[(otav[col] >= lower_bound) & (otav[col] <= upper_bound)]

    return df_otav_clean

columns = ['osrm_time', 'segment_osrm_time']
df_otav_outremoved = df_otav_remove_outliers(df_otav, columns)
df_otav_outremoved

In [ ]:
df_otav_outremoved.describe()

In [ ]:
t_statistic, p_value = stats.ttest_rel(df_otav_outremoved["osrm_time"], df_otav_outremoved["segment_osrm_time"])
print("t-statistic:", t_statistic)
print("p-value:", p_value)

alpha = 0.05
if p_value < alpha:
    print("Reject the null hypothesis. There is a significant difference between the means of osrm_time and segment_osrm_time.")
else:
    print("Fail to reject the null hypothesis. No significant difference between the means of osrm_time and segment_osrm_time.")

In [ ]:
df_melted_4 = pd.melt(df_otav_outremoved.reset_index(), id_vars = ["trip_uuid"], value_vars = ["osrm_time", "segment_osrm_time"])

sns.boxplot(x = "variable", y = "value", data = df_melted_4)

plt.title("Box plot for osrm time aggregated value and segment osrm time aggregated value")
plt.show()

### Hot encoding of categorical variables

In [ ]:
df_route_type = df.groupby("trip_uuid")["route_type"].first()
df_route_type.reset_index()
df_route_type

In [ ]:
df_route_type.value_counts()

In [ ]:
df_combined_times

In [ ]:
df_combined_times['route_type'] = df_combined_times['route_type'].map({'FTL':0, 'Carting':1})
df_combined_times.head()

### Normalize/ Standardize the numerical features using MinMaxScaler or StandardScaler

In [ ]:
df_scaled = df_combined_times.copy()
df_scaled = df_scaled.drop(columns = ["trip_uuid", "source_name", "destination_name"])
df_scaled

In [ ]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaled_data = scaler.fit_transform(df_scaled)
df_scaled_data = pd.DataFrame(scaled_data, columns = df_scaled.columns)
df_scaled_data.head()

In [ ]:
df_scaled_data.describe()